core/src/main/scala/Metrics.scala

```diff
@@ -168,7 +167,10 @@ object Metrics {
         _       <- onComplete.run(elapsed)
       } yield a
 
-      (io <* onSuccess).onError { case err => onFail(err) }
+      io.attempt.flatMap {
+        case Left(e)  => onFail(e) *> IO.raiseError(e)
+        case Right(a) => onSuccess *> IO.pure(a)
```

Comment:
Based on the original use of [`Task#onFinish`](https://github.com/scalaz/scalaz/blob/series/7.3.x/concurrent/src/main/scala/scalaz/concurrent/Task.scala#L50-L61) this seems to be a faithful translation, despite what the documentation of `onFinish` may suggest.